src/Core/src/Stream/ResponseBodyResourceStream.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\Core\Stream;
+
+use AsyncAws\Core\Exception\RuntimeException;
+
+/**
+ * Stream a resource body.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class ResponseBodyResourceStream implements ResultStream
+{
+    /**
+     * @var resource
+     */
+    private $resource;
+
+    public function __construct($resource)
+    {
+        $this->resource = $resource;
+    }
+
+    public function __toString()
+    {
+        return $this->getContentAsString();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getChunks(): iterable
+    {
+        $inPos = 0;
+        while (true) {
+            // move the cursor to reading position and remember the original position (it may changes)
+            $outPos = \ftell($this->resource);
+            if ($outPos !== $inPos && 0 !== \fseek($this->resource, $inPos)) {
+                throw new RuntimeException('The stream is not seekable');
+            }
+
+            $content = \fread($this->resource, 64 * 1024);
+            // break here as `feof` always returns false after calling `ftell`
+            if (\feof($this->resource)) {
+                if (\fseek($this->resource, $outPos)) {
+                    throw new RuntimeException('The stream is not seekable');
+                }
+                yield $content;
+
+                break;
+            }
+
+            // move the cursor to original position and remember the reading position
```

Comment:
If I read the stream, we are actually reading (or moving the cursor) the stream multiple times. 
If the stream in 10*64*1024 bytes (10 chunks). We are setting the cursor to the beginning 10 times. That is not needed. 


We dont need to reset the cursor on the "original position". We should make sure we read everything and don't care about other's reading the stream.

Reply:
Why not it would be much more easier.
but originally, my issue was: what if others trigger the code (HttpClient::steam) that write in the stream?

Reply:
nvm, the resource is used only by us. it should not be manipulated by somebody else